pulp_container/app/registry_api.py

```diff
@@ -554,12 +554,65 @@ def create(self, request, path):
         """
         _, repository = self.get_dr_push(request, path, create=True)
 
-        upload = models.Upload(repository=repository, size=0)
-        upload.save()
-        response = UploadResponse(upload=upload, path=path, request=request)
+        if self.tries_to_mount_blob(request):
+            response = self.mount_blob(request, path, repository)
+        else:
+            upload = models.Upload(repository=repository, size=0)
+            upload.save()
+            response = UploadResponse(upload=upload, path=path, request=request)
 
         return response
 
+    @staticmethod
+    def tries_to_mount_blob(request):
+        """Check if a client is trying to perform cross repository blob mounting."""
+        return (request.query_params.keys()) == {"from", "mount"}
+
+    def mount_blob(self, request, path, repository):
+        """Mount a blob that is already present in another repository."""
+        from_path = request.query_params["from"]
+        try:
+            distribution = models.ContainerDistribution.objects.get(base_path=from_path)
+        except models.ContainerDistribution.DoesNotExist:
+            raise RepositoryNotFound(name=path)
+
+        try:
+            version = distribution.repository_version or distribution.repository.latest_version()
+        except AttributeError:
+            # the distribution does not contain reference to repository version or repository
+            raise RepositoryNotFound(name=from_path)
+
+        digest = request.query_params["mount"]
+        try:
+            blob = models.Blob.objects.get(digest=digest, pk__in=version.content)
+        except models.Blob.DoesNotExist:
+            raise BlobNotFound(digest=digest)
+
+        dispatched_task = dispatch(
+            add_and_remove,
+            exclusive_resources=[version.repository, repository],
```

Comment:
you should put shared_lock on the repo you're importing the blob from. shared lock means one can perform read operations as  well as allowing others to perform read operations simultaneously. while exclusive lock does not allow other write not read operations.